packages/pocket-ic/src/lib.rs

```diff
@@ -649,6 +649,16 @@ impl PocketIc {
         runtime.block_on(async { self.pocket_ic.get_controllers(canister_id).await })
     }
 
+    /// Get the controllers of a canister.
+    #[instrument(ret, skip(self), fields(instance_id=self.pocket_ic.instance_id, canister_id = %canister_id.to_string()))]
+    pub fn try_get_controllers(
+        &self,
+        canister_id: CanisterId,
+    ) -> Result<Vec<Principal>, (StatusCode, String)> {
```

Comment:
Why not check if the canister exists via a separate call? Also returning the HTTP status code in this API looks strange.

Reply:
> Why not check if the canister exists via a separate call?

The behavior of `canister_info` which panics when the target canister doesn't exist was strange for me.
The panic is caused by the fact that `canister_info` used `get_controllers` which also panics when canister doesn't exist, so I decided to make it return some error instead of panicking

> Also returning the HTTP status code in this API looks strange

Agree

Reply:
> Also returning the HTTP status code in this API looks strange

Perhaps, the proper way to handle the error code is to assert that it is "not found" and return human-readable error via `String` and panic otherwise

Reply:
> to assert that it is "not found"

Disregard, the server returns `400 CanisterNotFound(CanisterId(rrkah-fqaaa-aaaaa-aaaaq-cai))` when canister doesn't exist

Reply:
I'd suggest to use `Option<Vec<Principal>>` as the return type and `None` means that the canister does not exist. You could also use [this](https://github.com/dfinity/ic/blob/edcb5ab41a0faa2801b93a9560b6f7f9a58b033d/packages/pocket-ic/src/lib.rs#L1131) function to check if a canister exists instead of relying on the HTTP status code.

Reply:
https://github.com/serokell/ic/pull/5/commits/bd419b509025dc68e523bc9f7d3fb53961e84d8a

This commit will land on the upstream sooner or later as a part of the 2nd milestone PR

Reply:
> You could also use [this](https://github.com/dfinity/ic/blob/edcb5ab41a0faa2801b93a9560b6f7f9a58b033d/packages/pocket-ic/src/lib.rs#L1131) function to check if a canister exists instead of relying on the HTTP status code.

Could you please also implement this suggestion?

Reply:
> This commit will land on the upstream sooner or later

Could you please merge it in the following days so that we can release PocketIC with the changes merged in?

Reply:
> Could you please also implement this suggestion?

Currently, my use case also includes checking that the canister was installed, so using `canister_info` makes slightly more sense because it returns WASM module hash

Reply:
> so using canister_info makes slightly more sense because it returns WASM module hash

how do you use `canister_info`? as anonymous read state requests analogously to how subnet metrics are fetched?

Reply:
> how do you use canister_info?

In my feature branch it is currently used the following way:
```
pub async fn check_canister_installed<C: CallCanisters>(
    agent: &C,
    canister_id: &CanisterId,
) -> bool {
    agent
        .canister_info(*canister_id)
        .await
        .map(|canister_info| canister_info.module_hash.is_some())
        .unwrap_or_default()
}
```

Reply:
I see: so you're using an agent to interact with a "live" PocketIC instance. Most PocketIC library APIs (such as `get_controllers` are supposed to work in the non-live mode where the instance does not necessarily make progress on its own).

Reply:
> I see: so you're using an agent to interact with a "live" PocketIC instance

Not necessarily, the goal is to be agnostic of client "backend" implementation (live/deterministic PocketIC or `ic-agent`)

Reply:
So how do you retrieve canister_info using a "deterministic" PocketIC?

Reply:
See [here](https://github.com/dfinity/ic/blob/master/rs/nervous_system/agent/src/pocketic_impl.rs#L145)

Reply:
https://github.com/dfinity/ic/pull/4026